src/mrack/providers/provider.py

```diff
@@ -106,34 +97,97 @@ async def provision_hosts(self, hosts):
+        missing_reqs = [req for req in reqs if req["name"] in error_host_names]
+        return (success_hosts, error_hosts, missing_reqs)
+
+    async def provision_hosts(self, reqs):
+        """Provision hosts based on list of host requirements.
+
+        Main provider method for provisioning.
 
-        if errors:
-            logger.info(f"{self.dsp_name}: Some host did not start properly")
-            for host_err in errors:
-                logger.error(f"{self.dsp_name}: Error: {str(host_err)}")
+        First it validates that host requirements are valid and that
+        provider has enough resources(quota).
 
-            logger.info(f"{self.dsp_name}: Given the error, will delete all hosts")
-            await self.delete_hosts(hosts)
-            raise ProvisioningError(errors)
+        Then issues provisioning and waits for it succeed. Raises exception if any of
+        the servers was not successfully provisioned. If that happens it issues deletion
+        of all already provisioned resources.
+
+        Return list of information about provisioned servers.
+        """
+        logger.info(f"{self.dsp_name}: Validating hosts definitions")
+        await self.validate_hosts(reqs)
+
+        if self.strategy == STRATEGY_RETRY:
+            success_hosts, error_hosts, missing_reqs = await self.strategy_retry(reqs)
+        else:
+            success_hosts, error_hosts, missing_reqs = await self.strategy_abort(reqs)
+
+        if error_hosts:
+            hosts_to_delete = success_hosts + error_hosts
+            self.abort_and_delete(hosts_to_delete, error_hosts)
 
         logger.info(f"{self.dsp_name}: Printing provisioned hosts")
-        for host in hosts:
+        for host in success_hosts:
             logger.info(f"{self.dsp_name}: {host}")
 
-        return hosts
-
-    def parse_errors(self, hosts):
+        return success_hosts
+
+    async def strategy_retry(self, reqs):
+        """Provisioning strategy to try multiple times to provision a host."""
+        missing_reqs = reqs
+        attempts = 0
+        success_hosts = []
+        error_hosts = []
+
+        while missing_reqs:
+            if attempts >= self.max_attempts:
+                logger.error(f"Max attempts({self.max_attempts}) reached. Aborting.")
+                break
+
+            attempts += 1
+            s_hosts, error_hosts, missing_reqs = await self._provision_base(
+                missing_reqs
+            )
+            success_hosts.extend(s_hosts)
+
+            if error_hosts:
+                count = len(error_hosts)
+                err = f"{count} hosts were not provisioned properly, deleting."
+                logger.info(f"{self.dsp_name}: {err}")
+                for host in error_hosts:
+                    logger.error(f"{self.dsp_name}: Error: {str(host.error)}")
+                await self.delete_hosts(error_hosts)
+                logger.info("Retrying to provision these hosts.")
```

Comment:
`"Retrying to provision these hosts. (Attempt {attempts} of {self.max_attempts}"` that might help us and users.